zebrad/src/components/sync/tests/vectors.rs

```diff
@@ -0,0 +1,134 @@
+    );
+
+    // This constraint avoids spurious failures due to block retries timing out.
+    // We multiply by 2, because the Hedge can wait up to BLOCK_DOWNLOAD_TIMEOUT
+    // seconds before retrying.
+    const BLOCK_DOWNLOAD_HEDGE_TIMEOUT: u64 =
+        2 * BLOCK_DOWNLOAD_RETRY_LIMIT as u64 * BLOCK_DOWNLOAD_TIMEOUT.as_secs();
+    assert!(
+        SYNC_RESTART_DELAY.as_secs() > BLOCK_DOWNLOAD_HEDGE_TIMEOUT,
+        "Sync restart should allow for block downloads to time out on every retry"
+    );
+
+    // This constraint avoids spurious failures due to block download timeouts
+    assert!(
+        BLOCK_VERIFY_TIMEOUT.as_secs()
+            > SYNC_RESTART_DELAY.as_secs()
+                + BLOCK_DOWNLOAD_HEDGE_TIMEOUT
+                + BLOCK_DOWNLOAD_TIMEOUT.as_secs(),
+        "Block verify should allow for a block timeout, a sync restart, and some block fetches"
+    );
+
+    // The minimum recommended network speed for Zebra, in bytes per second.
+    const MIN_NETWORK_SPEED_BYTES_PER_SEC: u64 = 10 * 1024 * 1024 / 8;
+
+    // This constraint avoids spurious failures when restarting large checkpoints
+    assert!(
+       BLOCK_VERIFY_TIMEOUT.as_secs() > SYNC_RESTART_DELAY.as_secs() + 2 * zebra_consensus::MAX_CHECKPOINT_BYTE_COUNT / MIN_NETWORK_SPEED_BYTES_PER_SEC,
+       "Block verify should allow for a full checkpoint download, a sync restart, then a full checkpoint re-download"
+    );
+
+    // This constraint avoids spurious failures after checkpointing has finished
+    assert!(
+        BLOCK_VERIFY_TIMEOUT.as_secs()
+            > 2 * zebra_chain::parameters::NetworkUpgrade::Blossom
+                .target_spacing()
+                .num_seconds() as u64,
+        "Block verify should allow for at least one new block to be generated and distributed"
+    );
+}
+
+/// Test that calls to [`ChainSync::request_genesis`] are rate limited.
+#[test]
+fn request_genesis_is_rate_limited() {
+    // create some counters that will be updated inside async blocks
+    let peer_requests_counter = Arc::new(AtomicU8::new(0));
+    let peer_requests_counter_clone = Arc::clone(&peer_requests_counter);
+    let state_requests_counter = Arc::new(AtomicU8::new(0));
+    let state_requests_counter_clone = Arc::clone(&state_requests_counter);
+    let verifier_requests_counter = Arc::new(AtomicU8::new(0));
+    let verifier_requests_counter_clone = Arc::clone(&verifier_requests_counter);
+
+    let runtime = Runtime::new().expect("Failed to create Tokio runtime");
+    let _guard = runtime.enter();
+
+    // create a fake peer service that respond with `Nil` to `BlocksByHash` or
+    // panic in any other type of request.
+    let peer_service = tower::service_fn(move |request| {
+        match request {
+            zebra_network::Request::BlocksByHash(_) => {
+                // Track the call
+                peer_requests_counter_clone.fetch_add(1, Ordering::SeqCst);
+                // Respond with `Nil`
```

Comment:
Return an error here instead.

That's what the downloaded expects when a block isn't available:
https://github.com/ZcashFoundation/zebra/pull/2255/files#diff-f11a3846db177395024935919250f553c3189810f934b10c9c12d4205ed2ed52L169